### ISSUE TITLE
Dataset is not unpickled correctly

### ISSUE THREAD

**Issue by iddan** [RDFLib/rdflib#893]
Dataset is unpickeld as ConjuctiveGraph because it extends ConjuctiveGraph and inherits the `__reduce__()` method. It should be unpickled as Dataset instead.

**Comment by iddan** [comment]
I've managed to fix it:
```python
from rdflib.graph import Dataset

class FixedDataset(Dataset):
    def __reduce__(self):
        return (type(self), (self.store, self.default_union))

    def __getstate__(self):
        return self.store, self.identifier, self.default_context, self.default_union

    def __setstate__(self, state):
        self.store, self.identifier, self.default_context, self.default_union = state

# TEST

import pickle
from rdflib.namespace import Namespace

example = Namespace("http://example.com#")

fixed = FixedDataset()
one = fixed.graph(example.one)

one.add((example.dan, example.knows, example.lisa))
one.add((example.lisa, example.knows, example.dan))

two = fixed.graph(example.two)

two.add((example.ben, example.knows, example.lisa))
two.add((example.lisa, example.knows, example.ben))

assert set(fixed.quads((None, None, None, None))) == set(fixed.quads((None, None, None, None)))

picklestring = pickle.dumps(fixed)
new_fixed = pickle.loads(picklestring)

for new_graph, graph in zip(new_fixed.graphs(), fixed.graphs()):
    assert new_graph.identifier == graph.identifier
    for new_triple, triple in zip(new_graph.triples((None, None, None)), graph.triples((None, None, None))):
        assert new_triple == triple
```

**Comment by gjhiggins** [comment]
Fixed in #1501
